airbyte-integrations/connectors/source-adjust/source_adjust/source.py

```diff
@@ -0,0 +1,211 @@
+
+import requests
+import source_adjust.model
+from airbyte_cdk.models import AirbyteMessage, AirbyteStateMessage, SyncMode, Type
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import IncrementalMixin, Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.requests_native_auth import TokenAuthenticator
+
+
+class AdjustStream(HttpStream, IncrementalMixin):
+    """
+    Adjust reports service integration with support for incremental synchronization.
+    """
+
+    def __init__(self, connector: "SourceAdjust", config: Mapping[str, Any], **kwargs):
+        super().__init__(**kwargs)
+
+        self.connector = connector
+        self.config = config
+        self._cursor: Optional[datetime.date] = None
+
+    @property
+    def url_base(self) -> str:
+        return "https://dash.adjust.com/control-center/reports-service/"
+
+    @property
+    def state(self):
+        if self._cursor is not None:
+            cursor = self._cursor.isoformat()
+        else:
+            cursor = self.config["ingest_start"]
+
+        return {
+            self.cursor_field: cursor,
+        }
+
+    @state.setter
+    def state(self, value: MutableMapping[str, Any]):
+        self._cursor = datetime.date.fromisoformat(value[self.cursor_field])
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: Optional[List[str]] = None,
+        stream_slice: Optional[Mapping[str, Any]] = None,
+        stream_state: Optional[Mapping[str, Any]] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        fallback = datetime.date.fromisoformat(self.config["ingest_start"])
+        cf: str = self.cursor_field
+
+        for record in super().read_records(sync_mode, cursor_field, stream_slice, stream_state):
+            record_stamp = datetime.date.fromisoformat(record[cf])
+            self._cursor = max(record_stamp, self._cursor or fallback)
+            yield record
+
+    def path(
+        self,
+        stream_state: Optional[Mapping[str, Any]] = None,
+        stream_slice: Optional[Mapping[str, Any]] = None,
+        next_page_token: Optional[Mapping[str, Any]] = None,
+    ) -> str:
+        """
+        Report URL path suffix.
+        """
+        return "report"
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Optional[Mapping[str, Any]] = None,
+        next_page_token: Optional[Mapping[str, Any]] = None,
+    ) -> MutableMapping[str, Any]:
+        """
+        Get query parameter definitions.
+        """
+        required_dimensions = ["day"]
+        dimensions = required_dimensions + self.config["dimensions"]
+        metrics = self.config["metrics"] + self.config["additional_metrics"]
+        date = stream_slice[self.cursor_field]
+        return {
+            "date_period": ":".join([date, date]),  # inclusive
```

Comment:
:+1: Did not test this. Could be an improvement.

Reply:
Considering this further however, it may be better leaving this as is. If the synch fails in the middle of a large ingest, we would have a checkpoint that could be restarted from.

Reply:
I noticed that the Adjust API supports pagination. We still are waiting on the credentials, which should be created soon. If this is the case, I would recommend using the `ingest_start:today` pattern. The Airbyte Python SDK already supports dealing with caching and pagination: https://docs.airbyte.com/connector-development/cdk-python/http-streams/.

Let me know what you think.